src/frequenz/sdk/timeseries/_moving_window.py

```diff
@@ -0,0 +1,170 @@
+
+    # ... wait for 5 minutes until the buffer is filled
+    await asyncio.sleep(5)
+
+    # return an numpy array from the window
+    a = window[time_start:time_end]
+    # and use it to for example calculate the mean
+    mean = a.mean()
+    '''
+
+    **Example2** (create a polars data frame from a `MovingWindow`):
+
+    ```
+    import polars as pl
+
+    # create a window that stores two days of data
+    # starting at 1.1.23 with samplerate=1
+    window = MovingWindow(size = (60 * 60 * 24 * 2), sample_receiver)
+
+    # wait for one full day until the buffer is filled
+    asyncio.sleep(60*60*24)
+
+    # create a polars series with one full day of data
+    time_start = datetime(2023, 1, 1)
+    time_end = datetime(2023, 1, 2)
+    s = pl.Series("Jan_1", mv[time_start:time_end])
+    ```
+    """
+
+    def __init__(
+        self,
+        size: int,
+        resampled_data_recv: Receiver[Sample],
+        sampling_period: timedelta,
+        window_alignment: datetime = datetime(1, 1, 1),
+    ) -> None:
+        """
+        Initialize the MovingWindow.
+
+        This method creates the underlying ringbuffer and starts a
+        new task that updates the ringbuffer with new incoming samples.
+        The task stops running only if the channel receiver is closed.
+
+        Args:
+            size: The number of elements that are stored.
+            resampled_data_recv: A receiver that delivers samples with a
+                given sampling period.
+            sampling_period: The sampling period.
+            window_alignment: A datetime object that defines a point in time to which
+                the window is aligned to modulo window size.
+                (default is midnight 01.01.01)
+
+        Raises:
+            asyncio.CancelledError: when the task gets cancelled.
+        """
+        self._resampled_data_recv = resampled_data_recv
+        self._buffer = OrderedRingBuffer(
+            np.empty(shape=size, dtype=float),
+            sampling_period=sampling_period,
+            time_index_alignment=window_alignment,
+        )
+        self._copy_buffer = False
+        try:
+            self._t: asyncio.Task = asyncio.create_task(self._run_impl())
+        except asyncio.CancelledError:
+            log.debug("Cancelling MovingWindow task: %s", __name__)
+            raise
+
+    async def _run_impl(self) -> None:
+        """Awaits samples from the receiver and updates the underlying ringbuffer."""
+        async for sample in self._resampled_data_recv:
+            log.debug("Received new sample: %s", sample)
+            self._buffer.update(sample)
+
+        log.error("Channel has been closed")
+
+    def __len__(self) -> int:
+        """
+        Return the size of the `MovingWindow`s underlying buffer.
+
+        Returns:
+            The size of the `MovingWindow`.
+        """
+        return len(self._buffer)
+
+    def __getitem__(self, key: Union[int, datetime, slice]) -> Union[float, ArrayLike]:
```

Comment:
```suggestion
    def __getitem__(self, key: int | datetime | slice) -> float | ArrayLike:
```

Reply:
That's what I did at first, but it doesn't work with python3.8.

Reply:
But we are using it!

https://github.com/frequenz-floss/frequenz-sdk-python/blob/v0.x.x/src/frequenz/sdk/timeseries/_ringbuffer.py#L389

Are you sure you tried it after you added the `from __future__ import annoations`?

Reply:
I suppose it doesn't work with more then two type operands.